src/license.rs

```diff
@@ -78,44 +81,75 @@ impl FromStr for License {
-            "GPL-3.0-only" | "GPL-3.0" => License::GPL_3_0,
-            "GPL-3.0-or-later" | "GPL-3.0+" => License::GPL_3_0Plus,
-            "AGPL-3.0-only" | "AGPL-3.0" => License::AGPL_3_0,
-            "AGPL-3.0-or-later" | "AGPL-3.0+" => License::AGPL_3_0Plus,
-            "Zlib" => License::Zlib,
-            // TODO: Sort out the SPDX "AND"
-            s if s.contains('/') || s.contains(" OR ") => {
-                let mut licenses = s
-                    .split('/')
-                    .flat_map(|s| s.split(" OR "))
-                    .map(str::parse)
-                    .map(Result::unwrap)
-                    .collect::<Vec<License>>();
-                licenses.sort();
-                License::Multiple(licenses)
+        if let Ok(expr) = spdx::expression::Expression::parse_mode(s, ParseMode::LAX) {
+            return Ok(process_spdx_expression(expr));
+        }
+        Ok(simple_license(s))
+    }
+}
+
+fn simple_license(s: &str) -> License {
+    match s.trim() {
+        "Unlicense" => License::Unlicense,
+        "0BSD" => License::BSD_0_Clause,
+        "CC0-1.0" => License::CC0_1_0,
+        "MIT" => License::MIT,
+        "X11" => License::X11,
+        "BSD-2-Clause" => License::BSD_2_Clause,
+        "BSD-3-Clause" => License::BSD_3_Clause,
+        "BSL-1.0" => License::BSL_1_0,
+        "Apache-2.0" => License::Apache_2_0,
+        "Apache-2.0 WITH LLVM-exception" => License::Apache_2_0_WITH_LLVM_exception,
+        "LGPL-2.0-only" | "LGPL-2.0" => License::LGPL_2_0,
+        "LGPL-2.1-only" | "LGPL-2.1" => License::LGPL_2_1,
+        "LGPL-2.1-or-later" | "LGPL-2.1+" => License::LGPL_2_1Plus,
+        "LGPL-3.0-only" | "LGPL-3.0" => License::LGPL_3_0,
+        "LGPL-3.0-or-later" | "LGPL-3.0+" => License::LGPL_3_0Plus,
+        "MPL-1.1" => License::MPL_1_1,
+        "MPL-2.0" => License::MPL_2_0,
+        "GPL-2.0-only" | "GPL-2.0" => License::GPL_2_0,
+        "GPL-2.0-or-later" | "GPL-2.0+" => License::GPL_2_0Plus,
+        "GPL-3.0-only" | "GPL-3.0" => License::GPL_3_0,
+        "GPL-3.0-or-later" | "GPL-3.0+" => License::GPL_3_0Plus,
+        "AGPL-3.0-only" | "AGPL-3.0" => License::AGPL_3_0,
+        "AGPL-3.0-or-later" | "AGPL-3.0+" => License::AGPL_3_0Plus,
+        "Zlib" => License::Zlib,
+        // TODO: Sort out the SPDX "AND"
+        s if s.contains('/') || s.contains(" OR ") => {
+            let mut licenses = s
+                .split('/')
+                .flat_map(|s| s.split(" OR "))
+                .map(str::parse)
+                .map(Result::unwrap)
+                .collect::<Vec<License>>();
+            licenses.sort();
+            License::Multiple(licenses)
+        }
+        s => License::Custom(s.to_owned()),
+    }
+}
+
+fn process_spdx_expression(expr: spdx::Expression) -> License {
+    let mut collection = Vec::new();
+    let mut queue = expr.iter().collect::<VecDeque<_>>();
+
+    while let Some(elem) = queue.pop_front() {
```

Comment:
Can you just iterate over the `expr` here instead of collecting into the `VecDeque`?

Also, the `collection` vec should be so small that it's probably easier to just check if an element already exists in it before pushing instead of dedupping with a hashset.

Reply:
I have done that now in 5d8b18145c970c375a127f4d21d25574d8e44288.